src/io/flutter/logging/FlutterLogEntryParser.java

```diff
@@ -63,33 +67,15 @@ static FlutterLogEntry parseDaemonEvent(@NotNull ProcessEvent event, @Nullable K
   @VisibleForTesting
   @Nullable
   public static FlutterLogEntry parseDaemonEvent(@NotNull String eventText) {
-    // Make a working copy.
-    String text = eventText;
-
-    // TODO(pq): replace heuristic parsing for JSON w/ something more robust.
-    if (text.startsWith("[")) {
-      // Remove bracketing braces.
-      text = text.substring(1, text.length() - 2);
-    }
-
-    try {
-      // TODO(pq): there's lots to parse; for now, just extract a message value.
-      final JsonObject json = new JsonParser().parse(text).getAsJsonObject();
-      final JsonObject params = json.get("params").getAsJsonObject();
-
-      if (params.has("message")) {
-        return new FlutterLogEntry(timestamp(), TOOLS_CATEGORY, params.get("message").getAsJsonPrimitive().getAsString());
+    stdoutParser.appendOutput(eventText);
+    for (String line : stdoutParser.getAvailableLines()) {
+      if (DaemonApi.parseAndValidateDaemonEvent(line.trim()) != null) {
```

Comment:
It seems like we are now parsing the DaemonEvent messages a couple times (here and separately when we actually do something with it). It would be cleaner to instead layer things so the DaemonEvent messages are only parsed once with daemon JSON going into one stream and regular log messages going somewhere else.

Reply:
Good idea.  Since that will likely touch a few other places, I'll add a TODO and loop back.